src/mkdocstrings/_cache.py

```diff
@@ -0,0 +1,76 @@
+import datetime
+import gzip
+import hashlib
+import os
+import urllib.parse
+import urllib.request
+from typing import BinaryIO, Callable
+
+import click
+import platformdirs
+
+from mkdocstrings.loggers import get_logger
+
+log = get_logger(__name__)
+
+
+def download_url_with_gz(url: str) -> bytes:
+    req = urllib.request.Request(  # noqa: S310
+        url,
+        headers={"Accept-Encoding": "gzip", "User-Agent": "mkdocstrings/0.15.0"},
+    )
+    with urllib.request.urlopen(req) as resp:  # noqa: S310
+        content: BinaryIO = resp
+        if "gzip" in resp.headers.get("content-encoding", ""):
+            content = gzip.GzipFile(fileobj=resp)  # type: ignore[assignment]
+        return content.read()
+
+
+# This is mostly a copy of https://github.com/mkdocs/mkdocs/blob/master/mkdocs/utils/cache.py
+# In the future maybe they can be deduplicated.
+
+
+def download_and_cache_url(
+    url: str,
+    download: Callable[[str], bytes],
+    cache_duration: datetime.timedelta,
+    comment: bytes = b"# ",
+) -> bytes:
+    """Downloads a file from the URL, stores it under ~/.cache/, and returns its content.
+
+    For tracking the age of the content, a prefix is inserted into the stored file, rather than relying on mtime.
+
+    Args:
+        url: URL to use.
+        download: Callback that will accept the URL and actually perform the download.
+        cache_duration: How long to consider the URL content cached.
+        comment: The appropriate comment prefix for this file format.
+    """
+    directory = os.path.join(platformdirs.user_cache_dir("mkdocs"), "mkdocstrings_url_cache")
+    name_hash = hashlib.sha256(url.encode()).hexdigest()[:32]
+    path = os.path.join(directory, name_hash + os.path.splitext(url)[1])
+
+    now = int(datetime.datetime.now(datetime.timezone.utc).timestamp())
+    prefix = b"%s%s downloaded at timestamp " % (comment, url.encode())
+    # Check for cached file and try to return it
+    if os.path.isfile(path):
+        try:
+            with open(path, "rb") as f:
+                line = f.readline()
+                if line.startswith(prefix):
+                    line = line[len(prefix) :]
+                    timestamp = int(line)
+                    if datetime.timedelta(seconds=(now - timestamp)) <= cache_duration:
+                        log.debug(f"Using cached '{path}' for '{url}'")
+                        return f.read()
+        except (OSError, ValueError) as e:
+            log.debug(f"{type(e).__name__}: {e}")
+
+    # Download and cache the file
+    log.debug(f"Downloading '{url}' to '{path}'")
+    content = download(url)
+    os.makedirs(directory, exist_ok=True)
+    with click.open_file(path, "wb", atomic=True) as f:
```

Comment:
So if I follow correctly, this function gets executed in a thread pool, and that's why we want atomic file writes. But in what circumstances could we have multiple writes on the same path? When the user is for example running `mkdocs build` and `mkdocs serve` at the same time :thinking: ? Or if they listed twice or more the same inventory in `mkdocs.yml`? In the worst case, the cache gets corrupted, but cache corruption seems to be handled above, in which case the file is simply re-downloaded. All in all, is the added `click` dependency worth it?

Reply:
> So if I follow correctly, this function gets executed in a thread pool, and that's why we want atomic file writes

No, this is if the write is interrupted in the middle and then you get a half-downloaded file cached permanently.

Reply:
OK, my main concern was the "added" Click dependency, but it's actually not added, as you explained in the other comment, so let's make good use of it.

Reply:
Ha. Now that I tried to actually replace it with something simpler, I ran into a race condition quite easily :smile: 

https://github.com/oprypin/mkdocs-gallery/actions/runs/6928061416/job/18843101055

mkdocs-gallery runs multiple `get-deps` at the same time

Reply:
My cross-platform, cross-language, dependency-free, dumb solution for file locking is to try and create directories as locks :rofl: Well I know it works on Linux because directory creation is atomic. I implement blocking (waiting) with a simple while loop + time sleep :shrug: